### PR TITLE
Add idp initiated login

### PR DESCRIPTION
**Pull request by mooreds**
Open questions:

On the API:

* do we handle starting a login? (I think no)
* can you complete a login using this idp (I think no, but the code said yes?)
* I used the same request/response fields with ifdef for the samlv2 and samlv2 idp initiated identity providers. Are these going to evolve together?

On the SAMLv2 IdP pages:

* I removed `issuer`, as it isn't there any longer in the UX. Not sure how to find out when it was removed to add a 'removed' notice.

For the SAMLv2 IdP initiated Identity Provider:

* There's no longer a CORS warning correct, because of the fix for https://github.com/FusionAuth/fusionauth-issues/issues/1203
* Do I overuse field and breadcrumb, esp for okta navigation? I wasn't quite sure how to handle that.

### CODE REVIEW

**Comment by robotdan** [comment]
> do we handle starting a login? (I think no)

Correct. IdP initiated means that this is an unsolicited request and we do know know ahead of time when a request will be sent. There is no "start" for this API. 

> can you complete a login using this idp (I think no, but the code said yes?)

Yes. Our front end always uses our own API. An integrator could use the `/api/identity-provider/login` API or when using our login pages we will call this API when the ACS (callback) is hit. 

> I used the same request/response fields with ifdef for the samlv2 and samlv2 idp initiated identity providers. Are these going to evolve together?

Unclear, they are discrete IdPs, there is some overlap. In theory they would evolve together, but much of the configuration is how to handle the request send to the IdP, and since that part is skipped in this case, it could just be that this config remains static. 

> I removed issuer, as it isn't there any longer in the UX. Not sure how to find out when it was removed to add a 'removed' notice.

I think this is ok, the `issuer` field is deprecated. FusionAuth generates this value and it is equal to `{public_url}/samlv2/sp/{identityProviderId}`. This value will be displayed in the UI view dialog. When we first shipped we allowed this to be set and quickly realized we had to set it. 

> There's no longer a CORS warning correct, because of the fix for FusionAuth/fusionauth-issues#1203

Correct. The end user will still likely need to configure CORS but because we cannot know that the `issues` is a URI, we cannot parse it to "guess" at what the correct configuration will be. We should still document in the UI walkthrough that you will likely need to configure an allowed origin for the requesting IdP so that you make it past CORS.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+Using IdP initiated SSO with this Identity Provider is not recommended and is inherently less secure than a service provider initiated login. 
+
+In general FusionAuth does not recommend you utilize this type of login due to the inherent risks of accepting an unsolicited authentication response from a third-party identity provider.  
+
+If you must enable this feature to support a legacy integration or because the user experience is of the highest concern, then proceed.
+====
+
+Typically, a SAML flow begins at a service provider such as FusionAuth. In a service provider initiated login, the SAML Authentication Response is returned to the service provider in response to an authentication request. FusionAuth can validate certain parameters and only accept a response for a request that FusionAuth initiated. 
+
+In contrast, during an identity provider initiated login, the third-party identity provider sends FusionAuth a SAML Authentication Response indicating a user has been authenticated without any context. FusionAuth did not initiate this request so it is unexpected. This limits what can be verified and requires FusionAuth to allow a cross site request from this third party.
```
Comment (robotdan) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:27]: >  has been authenticated without any context.

This was probably copied from me...  :-) but I wonder if we want a comma here to indicate that it wasn't that the user was "authenticated w/out context", but FusionAuth is sent a response w/out context, we didn't ask for it, we cannot be expecting it.

**Comment by mooreds** [comment]
Converting to draft to integrate Daniel's comments.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc (hunk 1 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:27]: makes sense. Leaving this comment open until approved so I can add a PR to change the text in the application UI as well.

File: site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc [hunk 2]
```diff
@@ -1,43 +1,72 @@
...
 [field]#identityProvider.enabled# [type]#[Boolean]# [optional]#Optional# [default]#defaults to `false`#::
 Determines if this provider is enabled. If it is false then it will be disabled globally.
 
+ifndef::samlv2_idp_initiated[]
 [field]#identityProvider.idpEndpoint# [type]#[String]# [optional]#Optional#::
 The SAML v2 login page of the identity provider.
+endif::[]
+
+ifdef::samlv2_idp_initiated[]
+[field]#identityProvider.issuer# [type]#[String]# [required]#Required#::
```
Comment (jobannon) [inline site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:67]: Do we have a flag that indicates `📣 Breaking Change - was required to be a URL until 1.27.1`(or something like that 🙂 )?

Probably not, but just checking.

File: site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc [hunk 3]
```diff
@@ -1,43 +1,72 @@
...
 
+ifndef::samlv2_idp_initiated[]
 [field]#identityProvider.debug# [type]#[Boolean]# [optional]#Optional# [default]#defaults to `false`# [since]#Available since 1.7.3#::
 Determines if debug is enabled for this provider. When enabled, each time this provider is invoked to reconcile a login an Event Log will be created.
 +
 This is useful for debugging purposes, but is not intended to be left on during production. It should be enabled during integration or if you are experiencing an issue with this identity provider.
+endif::[]
+ifdef::samlv2_idp_initiated[]
+[field]#identityProvider.debug# [type]#[Boolean]# [optional]#Optional# [default]#defaults to `false`#::
+Determines if debug is enabled for this provider. When enabled, each time this provider is invoked to reconcile a login an Event Log will be created.
```
Comment (jobannon) [inline site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:44]: consider
Determines if debug is enabled for this provider. When enabled, an Event Log is created each time this provider is invoked to reconcile a login.

File: site/docs/v1/tech/apis/identity-providers/_saml-complete-user-registration.adoc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+The user does not need to exist yet in FusionAuth to utilize this API. The `SAMLResponse` returned from the SAML v2 provider will be used to retrieve the user's email address, if that user does not yet exist in FusionAuth the user will be created. If [field]#Create registration# has been enabled for this identity provider and the user does not yet have a registration for this application, a registration will be automatically created for the user.
```
Comment (jobannon) [inline site/docs/v1/tech/apis/identity-providers/_saml-complete-user-registration.adoc]: consider 
email address, if => email address **_and_** if

Comment (jobannon) [reply site/docs/v1/tech/apis/identity-providers/_saml-complete-user-registration.adoc]: or consider
The user does not need to exist yet in FusionAuth to utilize this API. The `SAMLResponse` returned from the SAML v2 provider will be used to retrieve the user's email address.  FusionAuth if attempt to lookup by email and the user will be created if they do not yet exist.

File: site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc [hunk 5]
```diff
@@ -1,43 +1,72 @@
 ==== Request Body
```
Comment (bguyza) [inline site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:1]: Just a note that I verified optional/required fields in the code.  Looks good 👍

File: site/docs/v1/tech/apis/identity-providers/_samlv2-response-body.adoc [hunk 6]
```diff
@@ -37,12 +58,19 @@ Determines if this provider is enabled. If it is false then it will be disabled
...
+ifndef::samlv2_idp_initiated[]
 [field]#identityProvider.idpEndpoint# [type]#[String]#::
 The SAML v2 login page of the identity provider.
+endif::[]
 
 [field]#identityProvider.insertInstant# [type]#[Long]#::
 The link:/docs/v1/tech/reference/data-types/#instants[instant] that the provider was added to the FusionAuth database.
 
+ifdef::samlv2_idp_initiated[]
+[field]#identityProvider.issuer# [type]#[String]# [required]#Required#::
```
Comment (bguyza) [inline site/docs/v1/tech/apis/identity-providers/_samlv2-response-body.adoc:70]:
```suggestion
[field]#identityProvider.issuer# [type]#[String]#::
```
Omit required - this is a response body.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+In contrast, during an identity provider initiated login, the third-party identity provider sends FusionAuth a SAML Authentication Response indicating a user has been authenticated. FusionAuth did not initiate this request so it is unexpected and FusionAuth does not have the full context. This limits what can be verified and requires FusionAuth to allow a cross site request from this third party. 
+
+include::docs/v1/tech/shared/_premium-edition-blurb.adoc[]
+
+* <<Create a SAML v2 IdP Initiated Identity Provider>>
+* <<Integration Details>>
+* <<CORS Configuration>>
+
+We also provide specific examples for configuring SAML with some providers whose implementation requires unique configuration. If you'd like us to provide additional examples, please open a request on https://github.com/FusionAuth/fusionauth-site/issues[GitHub].
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:35]: consider
If you would like to see an example in addition to what has already been documented, please open a request on https://github.com/FusionAuth/fusionauth-site/issues[GitHub].

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+An optional UUID. When this value is omitted a unique Id will be generated automatically.
+
+[field]#Name# [required]#Required#::
+A unique name to identify the identity provider. This name is for display purposes only and it can be modified later if desired.
+
+[field]#Issuer# [required]#Required#::
+The EntityId (unique identifier) of the SAML v2 identity provider. This value should be provided to you.
+
+[field]#Use NameId for email# [optional]#Optional#::
+If this is enabled, FusionAuth will assume that the `NameID` in the SAML response contains the email address of the user.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:67]: consider
If this is enabled, FusionAuth will _**expect**_ the `NameID` in the SAML response contains the email address of the user.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+To configure a lambda, navigate to [breadcrumb]#Settings -> Lambdas#.
+
+[field]#Debug# [optional]#Optional#::
+Some identity providers are not compliant with the SAML and XML signing specifications. This makes it challenging to get them working with FusionAuth. 
++
+If you are running into integration issues, toggle this setting on and FusionAuth will output debugging information into the Event Log during a SAML login. You can find the event log in [breadcrumb]#System -> Event Log#.
+
+== Integration Details
+
+After configuring the Identity Provider, you can view the details to see values that will likely be required by your SAML v2 Identity Provider in order to trust FusionAuth as a relying party. Do so by navigating to [breadcrumb]#Settings -> Identity Providers# and click the green magnifying glass on your SAML provider.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:96]: Consider 
After configuring the Identity Provider, FusionAuth will display values likely required by your SAML v2 Identity Provider to trust FusionAuth as a relying party. Do so by navigating to [breadcrumb]#Settings -> Identity Providers# and click the green magnifying glass on your SAML provider.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
++
+If you are running into integration issues, toggle this setting on and FusionAuth will output debugging information into the Event Log during a SAML login. You can find the event log in [breadcrumb]#System -> Event Log#.
+
+== Integration Details
+
+After configuring the Identity Provider, you can view the details to see values that will likely be required by your SAML v2 Identity Provider in order to trust FusionAuth as a relying party. Do so by navigating to [breadcrumb]#Settings -> Identity Providers# and click the green magnifying glass on your SAML provider. 
+
+image::identity-providers/identity-provider-samlv2-idp-initiated-view-list.png[View the identity provider list,width=1200,role=bottom-cropped]
+
+When viewing the details, scroll to the [breadcrumb]#SAML v2 Integration details# section. There you will find the needed values to configure an integration with a SAMLv2 IdP.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:100]: small consideration 
needed => necessary

File: site/docs/v1/tech/apis/identity-providers/_saml-complete-user-registration.adoc (hunk 4 above)
Comment (bguyza) [inline site/docs/v1/tech/apis/identity-providers/_saml-complete-user-registration.adoc:1]:
```suggestion
The user does not need to exist yet in FusionAuth to utilize this API. The `SAMLResponse` returned from the SAML v2 provider will be used to retrieve the user's email address. If that user does not yet exist in FusionAuth the user will be created. If [field]#Create registration# has been enabled for this identity provider and the user does not yet have a registration for this application, a registration will be automatically created for the user. 
```

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+image::identity-providers/identity-provider-samlv2-idp-initiated-view-details.png[View the SAMLv2 identity provider details,width=1200]
+
+==== Fields 
+
+include::docs/v1/tech/identity-providers/_samlv2-integration-details-fields.adoc[]
+
+== CORS Configuration
+
+In order to complete the login request, the SAML v2 identity provider will make an HTTP `POST` request to the callback URL in FusionAuth. In order for this request to be allowed through the CORS filter you will need to navigate to [breadcrumb]#Settings -> System -> CORS# and add the SAML IdP origin as an [field]#Allowed Origin# the CORS configuration.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:112]: consider
~In order~ To complete the login request, the SAML v2 identity provider will make an HTTP `POST` request to the callback URL in FusionAuth. ~In order for this request to be allowed~ To allow this request through the CORS filter you will need to navigate to [breadcrumb]#Settings -> System -> CORS# and add the SAML IdP origin as an [field]#Allowed Origin# the CORS configuration.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+== Troubleshooting
+
+To troubleshoot, turn on the [field]#Debug# option and then navigate to [breadcrumb]#System -> Event Log#. 
+
+Proceed through the SAML flow and review the debug log entries to see if there are any configuration issues.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:118]: is it debug log or event log?

File: site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+== Overview
+
+[NOTE.since]
+====
+This API has been available since 1.26.0
+====
+
+include::docs/v1/tech/shared/_premium-edition-blurb.adoc[]
+
+The SAML v2 IdP Initiated IdP initiated Identity Provider allows an external IdP to send an unsolicited `AuthN` request when FusionAuth is acting as the Service Provider, or SP.
```
Comment (bguyza) [inline site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc:16]:
```suggestion
The SAML v2 IdP Initiated IdP initiated Identity Provider allows an external IdP to send an unsolicited `AuthN` request when FusionAuth is acting as the Service Provider (or SP).
```
I think putting the acronym in parenthesis reads better?  We do structure it this way on "docs/v1/tech/identity-providers/samlv2-idp-initiated/"

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+== Add the SAMLv2 IdP Initiated Identity Provider
+
+Navigate to [breadcrumb]#Settings -> Identity Providers#. Add a SAMLv2 IdP Initiated Provider.
+
+* Configure the [field]#Name# with a descriptive value. 
+* Set the [field]#Issuer# value to the [field]#Identity Provider Issuer# value from Okta.
+* Set the [field]#Verification key# to the public certificate you just imported.
+
+Enable this Identity Provider for any FusionAuth applications; for this document, `Pied Piper` allows use of this Identity Provider.
```
Comment (jobannon) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc:63]: consider 
Enable this Identity Provider for any FusionAuth applications you would like.  For this example, `Pied Piper` allows the use of this Identity Provider.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+[field]#Name# [required]#Required#::
+A unique name to identify the identity provider. This name is for display purposes only and it can be modified later if desired.
+
+[field]#Issuer# [required]#Required#::
+The EntityId (unique identifier) of the SAML v2 identity provider. This value should be provided to you.
+
+[field]#Use NameId for email# [optional]#Optional#::
+If this is enabled, FusionAuth will assume that the `NameID` in the SAML response contains the email address of the user.
+
+[field]#Email claim# [optional]#required# Required::
```
Comment (bguyza) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:69]:
```suggestion
[field]#Email claim# [required]#Required#::
```

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc [hunk 16]
```diff
@@ -0,0 +1,131 @@
+---
```
Comment (bguyza) [inline site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc:1]: I read through this page doing a once-over and looks good to me. I'd like to take the time to actually execute this example at some point, but I'm trusting the author has done this to confirm :)

**Review by bguyza — COMMENTED** [review]
Looks pretty good to me. I noted a few typos and/or recommended structure changes along the way.

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc (hunk 16 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/samlv2-idp-initiated/okta.adoc:1]: A few times, a few times. :)

**Review by jobannon — COMMENTED** [review]
Interesting to read about. .... minor stylistic suggestions

File: site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc [hunk 17]
```diff
@@ -1,43 +1,72 @@
...
+// no 'since' needed
+endif::[]
 
+ifndef::samlv2_idp_initiated[]
 [field]#identityProvider.domains# [type]#[Array<String>]# [optional]#Optional#::
 This is an optional list of domains that this SAML v2 provider should be used for. This converts the FusionAuth login form to a domain-based login form. This type of form first asks the user for their email. FusionAuth then uses their email to determine if an identity provider should be used. If an identity provider should be used, the browser is redirected to the login URL of that identity provider. Otherwise, the password field is revealed on the form so that the user can login using FusionAuth.
+endif::[]
 
 [field]#identityProvider.emailClaim# [type]#[String]# [optional]#Optional#::
 The name of the email claim (Attribute in the `Assertion` element) in the SAML response that FusionAuth uses to uniquely identity the user. If this is not set, the `useNameIdForEmail` flag must be true.
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:56]: `Attribute` is actually a SAML thing, I think.

[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:56]: I don't think that it is user uniqueness, I think it is just that we need to get the email from somewhere and if you don't tell us where, we'll assume it is in the `NameId` attribute.

File: site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc (hunk 2 above)
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/_samlv2-request-body.adoc:67]: No, we don't have that., and I'm not sure that is a breaking change. Going the other way would be, however (from requiring it to be a string to being a URL).

File: site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc (hunk 12 above)
Comment (mooreds) [reply site/docs/v1/tech/identity-providers/samlv2-idp-initiated/index.adoc:118]: Event log, thanks!

File: site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc]: haha!

`AuthN` is actually a SAML XML element, I believe.

File: site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+.URI
+--
+[method]#POST# [uri]#/api/identity-provider/login#
+--
+
+==== Request Headers
+
+[.api]
+[field]#X-Forwarded-For# [type]#[String]# [optional]#Optional#::
+The IP address of a client requesting authentication. If the IP address is provided it will be stored for login history of the user. It is generally preferred to specify the IP address in the request body. If it is not provided in the request body this header value will be used if available, the request body value will take precedence.
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc:200]: Thanks, this was helpful.

File: site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+* <<Complete a SAML v2 IdP Initiated Login>>
+
+=== Integration Details
+
+The following values will likely be required by your SAML v2 IdP Initiated Identity Provider in order to trust FusionAuth as a relying party.
+
+They are autogenerated and viewable within the UI after creating the Identity Provider. [breadcrumb]#Settings -> Identity Providers -> SAMLv2 IdP Initiated -> View#
+
+[cols="1,3"]
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/apis/identity-providers/samlv2-idp-initiated.adoc:35]: Not sure, probably a matter of pref

**Pull request merged**